File: mu/clients/mu-client.py
#!/usr/bin/env python3
"""Приклад клієнта для unix socket API μ-daemon.

Надсилає платіжний intent, читає відповідь.

Використання:
  python3 mu-client.py <socket_path> <recipient> <amount_minor> <agent_id>

Приклад:
  python3 mu-client.py /tmp/mu-policy.sock 0x8335... 5000000 agent-1
"""

import socket, struct, sys


def build_frame(recipient: str, amount: int, chain_id: int,
                agent_id: str, nonce: int, ts: int) -> bytes:
    """Збирає кадр у форматі mu-gate (wire.rs)."""
    payload = bytearray()
    payload += struct.pack('>H', 1)                    # version
    payload += struct.pack('B', len(recipient))        # recipient len
    payload += recipient.encode('ascii')
    payload += struct.pack('>Q', amount >> 64)          # amount u128 BE
    payload += struct.pack('>Q', amount & 0xFFFFFFFFFFFFFFFF)
    payload += struct.pack('>Q', chain_id)
    payload += struct.pack('B', len(agent_id))
    payload += agent_id.encode('ascii')
    payload += struct.pack('>Q', nonce)
    payload += struct.pack('>Q', ts)
    payload += b'\x00' * 64                             # підпис (тест)
    frame = struct.pack('<I', len(payload)) + payload
    return bytes(frame)


def main():
    if len(sys.argv) < 5:
        print(__doc__)
        sys.exit(1)

    sock_path = sys.argv[1]
    recipient = sys.argv[2]
    amount = int(sys.argv[3])
    agent_id = sys.argv[4]

    frame = build_frame(recipient, amount, 8453, agent_id, nonce=1, ts=1000)

    sock = socket.socket(socket.AF_UNIX, socket.SOCK_STREAM)
    sock.connect(sock_path)
    sock.sendall(frame)
    resp = sock.recv(1)
    sock.close()

    codes = {
        0x00: 'Parse error',
        0x01: 'Unknown agent',
        0x02: 'Bad signature',
        0x03: 'Stale timestamp',
        0x04: 'Replay nonce',
        0x05: 'Rate limited',
        0x06: 'Busy',
    }
    code = resp[0]
    if code == 0xFF:
        print('OK: intent accepted')
    else:
        print(f'Denied: {codes.get(code, f"unknown {code:#x}")}')


if __name__ == '__main__':
    main()

File: mu/clients/mu-client.ts
/**
 * Приклад клієнта для unix socket API μ-daemon (TypeScript/Node.js).
 *
 * Використання:
 *   npx ts-node mu-client.ts <socket_path> <recipient> <amount_minor> <agent_id>
 *
 * Залежності: Node.js 18+ з net module.
 */

import * as net from 'net';
import * as process from 'process';

function buildFrame(
  recipient: string,
  amount: bigint,
  chainId: number,
  agentId: string,
  nonce: number,
  ts: number,
): Buffer {
  const payload = Buffer.alloc(4096);
  let off = 0;

  // version u16 BE
  payload.writeUInt16BE(1, off); off += 2;
  // recipient len u8 + data
  payload[off++] = recipient.length;
  payload.write(recipient, off, recipient.length, 'ascii'); off += recipient.length;
  // amount u128 BE (hi+lo)
  payload.writeBigUInt64BE(amount >> 64n, off); off += 8;
  payload.writeBigUInt64BE(amount & 0xFFFFFFFFFFFFFFFFn, off); off += 8;
  // chain_id u64 BE
  payload.writeBigUInt64BE(BigInt(chainId), off); off += 8;
  // agent_id len + data
  payload[off++] = agentId.length;
  payload.write(agentId, off, agentId.length, 'ascii'); off += agentId.length;
  // nonce u64 BE
  payload.writeBigUInt64BE(BigInt(nonce), off); off += 8;
  // ts u64 BE
  payload.writeBigUInt64BE(BigInt(ts), off); off += 8;
  // signature (test/fake)
  off += 64;

  const body = payload.subarray(0, off);
  const header = Buffer.alloc(4);
  header.writeUInt32LE(body.length, 0);
  return Buffer.concat([header, body]);
}

function main() {
  const args = process.argv.slice(2);
  if (args.length < 4) {
    console.log(`Usage: ts-node mu-client.ts <socket> <recipient> <amount> <agent_id>`);
    process.exit(1);
  }

  const [sockPath, recipient, amountStr, agentId] = args;
  const amount = BigInt(amountStr);

  const frame = buildFrame(recipient, amount, 8453, agentId, 1, 1000);
  const sock = net.createConnection(sockPath);

  sock.on('connect', () => {
    sock.write(frame);
  });

  sock.on('data', (data: Buffer) => {
    const code = data[0];
    const codes: Record<number, string> = {
      0x00: 'Parse error',
      0x01: 'Unknown agent',
      0x02: 'Bad signature',
      0x03: 'Stale timestamp',
      0x04: 'Replay nonce',
      0x05: 'Rate limited',
      0x06: 'Busy',
    };
    if (code === 0xFF) {
      console.log('OK: intent accepted');
    } else {
      console.log(`Denied: ${codes[code] ?? `unknown ${'0x' + code.toString(16)}`}`);
    }
    sock.end();
  });

  sock.on('error', (err) => {
    console.error('Socket error:', err.message);
    process.exit(1);
  });
}

main();
